refactor: Standardize StrategyV1 and Adapters, rename Azorius to ProposalInitializer

adamgall opened the pull request:
Fixes [ENG-1057](https://linear.app/decent-labs/issue/ENG-1057/refactor-strategyv1-and-adapters-rename-azorius-and-remove-initialize)

This commit refactors the `StrategyV1` contract and its various ERC20, ERC721, and Hats Proposer/Voting adapters. The key changes include:

1.  **Renaming `azorius` to `proposalInitializer`:**
    *   In `StrategyV1.sol`, the internal state variable `_azorius` has been renamed to `_proposalInitializer`.
    *   The public getter `azorius()` has been renamed to `proposalInitializer()`.
    *   The `initialize` function parameter `azorius_` has been renamed to `proposalInitializer_`.
    *   The corresponding `IStrategyV1` interface has been updated with the new function name `proposalInitializer()`.
    *   The `InvalidAzoriusAddress` error has been replaced with `InvalidProposalInitializer` in the interface.

2.  **Introduction of `onlyProposalInitializer` Modifier:**
    *   A new modifier `onlyProposalInitializer` has been added to `StrategyV1.sol`. This modifier restricts access to functions to only the `_proposalInitializer` address.
    *   The `initializeProposal` function in `StrategyV1.sol` now uses this `onlyProposalInitializer` modifier instead of a direct `msg.sender != _azorius` check.

3.  **Removal of Initialization Checks and Associated Errors/Events:**
    *   **`StrategyV1.sol`:**
        *   Removed checks from `initialize` for:
            *   `azorius_` (now `proposalInitializer_`) being the zero address.
            *   `votingPeriod_` being zero.
            *   `votingAdapters_` containing zero addresses or duplicates.
            *   `proposerAdapters_` containing zero addresses or duplicates.
        *   Removed corresponding errors from `IStrategyV1.sol`: `InvalidAzoriusAddress` (replaced), `InvalidVotingPeriod`, `VotingAdapterIsZeroAddress`, `VotingAdapterAlreadyExists`, `NoVotingAdapters`, `ProposerAdapterIsZeroAddress`, `ProposerAdapterAlreadyExists`.
    *   **Adapter Contracts (`ERC20ProposerAdapterV1`, `ERC20VotingAdapterV1`, `ERC721ProposerAdapterV1`, `ERC721VotingAdapterV1`, `HatsProposerAdapterV1`):**
        *   Removed checks from their respective `initialize` functions for:
            *   Token addresses being zero (`InvalidTokenAddress` error).
            *   Strategy address being zero (`InvalidStrategyAddress` error in `ERC20VotingAdapterV1`).
            *   `weightPerToken_` being zero (`InvalidWeightPerToken` error).
            *   `hatsContract_` being zero (`MissingHatsContract` error in `HatsProposerAdapterV1`).
            *   `whitelistedHatIds_` being empty (`NoHatsWhitelisted` error in `HatsProposerAdapterV1`).
            *   Duplicate hat IDs in `whitelistedHatIds_` (`HatAlreadyWhitelisted` error in `HatsProposerAdapterV1`).
        *   Removed corresponding error definitions from their respective interfaces (`IERC20ProposerAdapterV1`, `IERC20VotingAdapterV1`, etc.).

4.  **Test Updates:**
    *   **`StrategyV1.test.ts`**:
        *   Updated `azoriusMock` to `proposerInitializer` and related address usage.
        *   Removed tests for initialization checks that were removed from `StrategyV1.initialize`.
        *   Updated the `initializeProposal` test to reflect the usage of `onlyProposalInitializer` modifier.
    *   **Adapter Test Files (`ERC20ProposerAdapterV1.test.ts`, etc.)**:
        *   Removed tests for initialization checks that were removed from the adapter `initialize` functions.

**Rationale for Changes:**

*   The renaming from `azorius` to `proposalInitializer` provides more clarity on the role of this address – it's the entity authorized to initialize proposals within the strategy, which might not always be an Azorius module itself.
*   The `onlyProposalInitializer` modifier centralizes access control for proposal initialization.
*   Removing extensive input validation from `initialize` functions can simplify the contracts and potentially save gas on deployment. It shifts the responsibility for providing valid parameters to the deployer or factory contract. This is a common pattern where complex off-chain validation is preferred over on-chain checks for deployment parameters.

These changes streamline the `StrategyV1` and its adapter contracts, making them more focused and relying on the deploying entity for parameter correctness.

adamgall committed:
refactor: Standardize StrategyV1 and Adapters, rename Azorius to ProposalInitializer

This commit refactors the `StrategyV1` contract and its various ERC20, ERC721, and Hats Proposer/Voting adapters. The key changes include:

1.  **Renaming `azorius` to `proposalInitializer`:**
    *   In `StrategyV1.sol`, the internal state variable `_azorius` has been renamed to `_proposalInitializer`.
    *   The public getter `azorius()` has been renamed to `proposalInitializer()`.
    *   The `initialize` function parameter `azorius_` has been renamed to `proposalInitializer_`.
    *   The corresponding `IStrategyV1` interface has been updated with the new function name `proposalInitializer()`.
    *   The `InvalidAzoriusAddress` error has been replaced with `InvalidProposalInitializer` in the interface.

2.  **Introduction of `onlyProposalInitializer` Modifier:**
    *   A new modifier `onlyProposalInitializer` has been added to `StrategyV1.sol`. This modifier restricts access to functions to only the `_proposalInitializer` address.
    *   The `initializeProposal` function in `StrategyV1.sol` now uses this `onlyProposalInitializer` modifier instead of a direct `msg.sender != _azorius` check.

3.  **Removal of Initialization Checks and Associated Errors/Events:**
    *   **`StrategyV1.sol`:**
        *   Removed checks from `initialize` for:
            *   `azorius_` (now `proposalInitializer_`) being the zero address.
            *   `votingPeriod_` being zero.
            *   `basisNumerator_` validity (being within `BASIS_DENOMINATOR / 2` and `< BASIS_DENOMINATOR`).
            *   `votingAdapters_` containing zero addresses or duplicates.
            *   `proposerAdapters_` containing zero addresses or duplicates.
        *   Removed corresponding errors from `IStrategyV1.sol`: `InvalidAzoriusAddress` (replaced), `InvalidVotingPeriod`, `InvalidBasisNumerator`, `VotingAdapterIsZeroAddress`, `VotingAdapterAlreadyExists`, `NoVotingAdapters`, `ProposerAdapterIsZeroAddress`, `ProposerAdapterAlreadyExists`.
    *   **Adapter Contracts (`ERC20ProposerAdapterV1`, `ERC20VotingAdapterV1`, `ERC721ProposerAdapterV1`, `ERC721VotingAdapterV1`, `HatsProposerAdapterV1`):**
        *   Removed checks from their respective `initialize` functions for:
            *   Token addresses being zero (`InvalidTokenAddress` error).
            *   Strategy address being zero (`InvalidStrategyAddress` error in `ERC20VotingAdapterV1`).
            *   `weightPerToken_` being zero (`InvalidWeightPerToken` error).
            *   `hatsContract_` being zero (`MissingHatsContract` error in `HatsProposerAdapterV1`).
            *   `whitelistedHatIds_` being empty (`NoHatsWhitelisted` error in `HatsProposerAdapterV1`).
            *   Duplicate hat IDs in `whitelistedHatIds_` (`HatAlreadyWhitelisted` error in `HatsProposerAdapterV1`).
        *   Removed corresponding error definitions from their respective interfaces (`IERC20ProposerAdapterV1`, `IERC20VotingAdapterV1`, etc.).

4.  **Test Updates:**
    *   **`StrategyV1.test.ts`**:
        *   Updated `azoriusMock` to `proposerInitializer` and related address usage.
        *   Removed tests for initialization checks that were removed from `StrategyV1.initialize`.
        *   Updated the `initializeProposal` test to reflect the usage of `onlyProposalInitializer` modifier.
    *   **Adapter Test Files (`ERC20ProposerAdapterV1.test.ts`, etc.)**:
        *   Removed tests for initialization checks that were removed from the adapter `initialize` functions.

**Rationale for Changes:**

*   The renaming from `azorius` to `proposalInitializer` provides more clarity on the role of this address – it's the entity authorized to initialize proposals within the strategy, which might not always be an Azorius module itself.
*   The `onlyProposalInitializer` modifier centralizes access control for proposal initialization.
*   Removing extensive input validation from `initialize` functions can simplify the contracts and potentially save gas on deployment. It shifts the responsibility for providing valid parameters to the deployer or factory contract. This is a common pattern where complex off-chain validation is preferred over on-chain checks for deployment parameters.

These changes streamline the `StrategyV1` and its adapter contracts, making them more focused and relying on the deploying entity for parameter correctness.

tbwebb22 commented on contracts/deployables/strategies/StrategyV1.sol:
```diff
@@ -29,59 +29,42 @@ contract StrategyV1 is
-
         _quorumThreshold = quorumThreshold_;
-
-        if (
-            basisNumerator_ >= BASIS_DENOMINATOR ||
-            basisNumerator_ < BASIS_DENOMINATOR / 2
-        ) revert InvalidBasisNumerator();
```
I know we talked about removing initializer sanity checks - this is one I could see keeping since its enforcing a mathematical relationship

adamgall replied:
great point, will put it back

adamgall replied:
pushed, please re-review.

https://github.com/decentdao/decent-contracts/pull/307/commits/fb0f765fe5d79509ae26debdbb373f18c3d05c7c

tbwebb22 commented on contracts/deployables/strategies/adapters/HatsProposerAdapterV1.sol:
```diff
@@ -28,18 +28,8 @@ contract HatsProposerAdapterV1 is
-            }
-            _whitelistedHatIds.push(_hatId);
-        }
+        _whitelistedHatIds = whitelistedHatIds_;
```
so with this change a hatId could be whitelisted twice? Are we alright with that?

adamgall replied:
it doesn't hurt anything. i'm ok with it.

adamgall committed:
feat: Add validation for basis numerator in StrategyV1 initialization

This commit introduces a check in the `initialize` function of the `StrategyV1` contract to ensure that the `basisNumerator_` is within valid bounds (greater than or equal to 50% and less than 100%). An `InvalidBasisNumerator` error is defined in the `IStrategyV1` interface to handle invalid cases. Additionally, new tests are added to verify the correct behavior when initializing with valid and invalid basis numerators, ensuring robust contract initialization.
